engine/class_modules/paladin/sc_paladin_retribution.cpp

```diff
@@ -957,37 +960,33 @@ void paladin_t::generate_action_prio_list_ret()
-  generators -> add_action( this, "Hammer of Wrath", "if=holy_power<=4" );
+  finishers -> add_talent( this, "Execution Sentence", "if=((!talent.crusade.enabled|buff.crusade.down&cooldown.crusade.remains>10)|buff.crusade.stack>=3|cooldown.avenging_wrath.remains>10|debuff.final_reckoning.up)&time_to_hpg=0&target.time_to_die>8" );
+  finishers -> add_action( this, "Divine Storm", "if=variable.ds_castable&!buff.vanquishers_hammer.up&((!talent.crusade.enabled|cooldown.crusade.remains>gcd*3)&(!talent.execution_sentence.enabled|(cooldown.execution_sentence.remains%gcd+holy_power>7|target.time_to_die<8))&(!talent.final_reckoning.enabled|cooldown.final_reckoning.remains%gcd+holy_power>7)|spell_targets.divine_storm>=2&(talent.holy_avenger.enabled&cooldown.holy_avenger.remains<gcd*3|buff.crusade.up&buff.crusade.stack<10))" );
+  finishers -> add_action( this, "Templar's Verdict", "if=(!talent.crusade.enabled|cooldown.crusade.remains>gcd*3)&(!talent.execution_sentence.enabled|(cooldown.execution_sentence.remains%gcd+holy_power>7|target.time_to_die<8))&(!talent.final_reckoning.enabled|cooldown.final_reckoning.remains%gcd+holy_power>7)&(!covenant.necrolord.enabled|cooldown.vanquishers_hammer.remains>gcd)|talent.holy_avenger.enabled&cooldown.holy_avenger.remains<gcd*3|buff.holy_avenger.up|buff.crusade.up&buff.crusade.stack<10|buff.vanquishers_hammer.up" );
+
+  generators -> add_action( "call_action_list,name=finishers,if=holy_power=5|buff.holy_avenger.up|debuff.final_reckoning.up|debuff.execution_sentence.up" );
+  generators -> add_action( "divine_toll,if=!debuff.judgment.up&(!raid_event.adds.exists|raid_event.adds.in>30)&(holy_power<=2|holy_power<=4&(cooldown.blade_of_justice.remains>gcd*2|(debuff.execution_sentence.up|target.time_to_die<8)|debuff.final_reckoning.up))&(!talent.final_reckoning.enabled|cooldown.final_reckoning.remains>gcd*10)&(!talent.execution_sentence.enabled|(cooldown.execution_sentence.remains>gcd*10|target.time_to_die<8))&(cooldown.avenging_wrath.remains>0|cooldown.crusade.remains>0)" );
+  generators -> add_action( this, "Wake of Ashes", "if=(holy_power=0|holy_power<=2&(cooldown.blade_of_justice.remains>gcd*2|(debuff.execution_sentence.up|target.time_to_die<8)|debuff.final_reckoning.up))&(!raid_event.adds.exists|raid_event.adds.in>20)&(!talent.execution_sentence.enabled|cooldown.execution_sentence.remains>15)&(!talent.final_reckoning.enabled|cooldown.final_reckoning.remains>15)&(cooldown.avenging_wrath.remains>0|cooldown.crusade.remains>0)" );
+  generators -> add_action( this, "Blade of Justice", "if=holy_power<=3&talent.execution_sentence.enabled&(debuff.execution_sentence.up|cooldown.execution_sentence.remains=0)" );
+  generators -> add_action( this, "Hammer of Wrath" );
```

Comment:
it's somewhat surprising that just dropping the clash-avoidance and waste-avoidance logic on these lines leads to an improvement. How well does this work across talent setups that don't involve ES, in particular TFoJ/DP or BoW/DP combinations?

Reply:
The check for BoJ not being up within 2 GCDs to use Judgment is performing lower for every talent combination I tried including those two, I suspect because the sequence of TV - BoJ - Judge just puts your generators on cooldown slower than Judge - TV - BoJ, even if BoJ is a higher priority generator.

I removed the checks for Holy Power from abilities that generate 1 Holy Power just because they're redundant, since it'll just use a spender instead of a generator above 4 Holy Power anyway. It doesn't make any actual performance difference, I can put them back in if they're there to make it read more intuitively.